# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, you'll gain practice combining DataFrames through concatenation.  You'll also practice executing various types of joins to selectively combine the information stored in the tables!

## Objectives

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)


## Getting Started

You'll start with a quick section to gain practice with concatenating datasets using `pd.concat()`.

## Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [1]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

Now that you have multiple DataFrames to work with, you can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.  

In [2]:
combined_df = pd.concat([df1, df2, df3])
combined_df

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

**_EXPECTED OUTPUT:_**

<img src="images/er1.png">

## Setting Join Conditions With Concatenation

You can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, you'll execute an inner join. Before you do, you need to create another table that contains some overlapping index values with a DataFrame that already exists.  

Run the cell below to create the new DataFrame.

In [3]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7])

Now, in the cell below, use the `pd.concat()` method to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [7]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [8]:
df4

B   D   F
2  B2  D2  F2
3  B3  D3  F3
6  B6  D6  F6
7  B7  D7  F7

In [11]:
df1_and_4 = pd.concat([df1, df4], join='inner', axis=1)
df1_and_4

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**_EXPECTED OUTPUT:_**

<img src='images/er2.png'>

You'll notice that in this case, the results returned contain only the rows with indexes that exist in both tables--rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that you can make full use of the `concat()` functionality in pandas to join DataFrames together--these are just a few of the most common examples pulled from the pandas documentation on the subject.  For a full view of all the ways you can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Loading In Data
Now, it's time to move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the  [kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table--but not all--correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  You'll make use of these keys to **_join_** the different tables together into a single DataFrame.  You'll also experiment with different types of joins to help us decide exactly what information you wish to combine.  

Simply run the cell below to import the tables from the database as DataFrames.

In [12]:
cards_df = pd.read_csv('cards.csv')
dust_df = pd.read_csv('dust.csv')
entourages_df = pd.read_csv('entourages.csv')
mechanics_df = pd.read_csv('mechanics.csv')
play_requirements_df = pd.read_csv('play_requirements.csv')

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the dataframes you created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the head of each respective DataFrame to ensure everything worked.  

**_NOTE:_** Since you are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

In [17]:
cards_df.reset_index(inplace=True)
cards_df.set_index('card_id', inplace=True)
cards_df.head()

player_class    type                name      set  \
card_id                                                        
KARA_00_07         MAGE   SPELL       Astral Portal     KARA   
NEW1_008a         DRUID   SPELL   Ancient Teachings  EXPERT1   
BRM_010t2         DRUID  MINION  Druid of the Flame      BRM   
AT_132          NEUTRAL  MINION  Justicar Trueheart      TGT   
OG_141          NEUTRAL  MINION   Faceless Behemoth       OG   

                                                         text  cost  attack  \
card_id                                                                       
KARA_00_07           Summon a random <b>Legendary</b> minion.   1.0     NaN   
NEW1_008a                                        Draw a card.   0.0     NaN   
BRM_010t2                                                 NaN   3.0     2.0   
AT_132      <b>Battlecry:</b> Replace your starting Hero P...   6.0     6.0   
OG_141                                                    NaN  10.0    10.0   

            health     rarity  collectible  \
card_id                                      
KARA_00_07     NaN        NaN          NaN   
NEW1_008a      NaN        NaN          NaN   
BRM_010t2      5.0     COMMON          NaN   
AT_132         3.0  LEGENDARY          1.0   
OG_141        10.0     COMMON          1.0   

                                                       flavor   race  \
card_id                                                                
KARA_00_07                                                NaN    NaN   
NEW1_008a                                                 NaN    NaN   
BRM_010t2                                                 NaN  BEAST   
AT_132      It's like putting racing stripes and a giant s...    NaN   
OG_141      Rejected names: Forty-Foot Faceless, Big ol' N...    NaN   

           how_to_earn how_to_earn_golden targeting_arrow_text faction  \
card_id                                                                  
KARA_00_07         NaN                NaN                  NaN     NaN   
NEW1_008a          NaN                NaN                  NaN     NaN   
BRM_010t2          NaN                NaN                  NaN     NaN   
AT_132             NaN                NaN                  NaN     NaN   
OG_141             NaN                NaN                  NaN     NaN   

            durability  
card_id                 
KARA_00_07         NaN  
NEW1_008a          NaN  
BRM_010t2          NaN  
AT_132             NaN  
OG_141             NaN

In [18]:
dust_df.set_index('card_id', inplace=True)
dust_df.head()

action  cost
card_id                           
BRM_010t2    CRAFTING_NORMAL    40
BRM_010t2    CRAFTING_GOLDEN   400
BRM_010t2  DISENCHANT_NORMAL     5
BRM_010t2  DISENCHANT_GOLDEN    50
AT_132       CRAFTING_NORMAL  1600

In [21]:
entourages_df.head()

card_id entourage_card_id
0  KAR_A10_22        KAR_A10_09
1  KAR_A10_22        KAR_A10_02
2  KAR_A10_22        KAR_A10_08
3  KAR_A10_22        KAR_A10_04
4  KAR_A10_22        KAR_A10_05

In [22]:
mechanics_df.set_index('card_id', inplace=True)

In [23]:
mechanics_df.head()

mechanic
card_id                      
AT_132              BATTLECRY
GVG_011a  TAG_ONE_TURN_EFFECT
EX1_583             BATTLECRY
LOE_007t            EVIL_GLOW
LOE_007t   ImmuneToSpellpower

In [24]:
play_requirements_df.set_index('card_id', inplace=True)

In [25]:
play_requirements_df.head()

play_requirement  value
card_id                                     
KARA_00_07       REQ_NUM_MINION_SLOTS      1
PRO_001a         REQ_NUM_MINION_SLOTS      1
NAX1_01          REQ_NUM_MINION_SLOTS      1
DS1h_292_H1           REQ_STEADY_SHOT      0
DS1h_292_H1  REQ_MINION_OR_ENEMY_HERO      0

## Executing Joins

Now that you have the tables loaded correctly, we're going to execute some joins. There are 4 different kinds of joins, which can best be visualized with venn diagrams:

<img src='images/Image_198_joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table.  The left table is the table you are working with, and the right table is the table you want to join to the table you are working with.  You'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table you want to join with, `mechanics_df`
* The `how` parameter set to the type of join you want, `'inner'`

In [27]:
cards_with_mechanics_df = cards_df.join(mechanics_df, on='card_id', how='inner')
print(cards_df.shape[0])
print(mechanics_df.shape[0])
print(cards_with_mechanics_df.shape[0])
cards_with_mechanics_df

2819
1079
1079


player_class         type                   name      set  \
card_id                                                                        
AT_132                  NEUTRAL       MINION     Justicar Trueheart      TGT   
GVG_011a                NEUTRAL  ENCHANTMENT             Shrink Ray      GVG   
EX1_583                 NEUTRAL       MINION     Priestess of Elune  EXPERT1   
LOE_007t                WARLOCK        SPELL                Cursed!      LOE   
LOE_007t                WARLOCK        SPELL                Cursed!      LOE   
CFM_095                 NEUTRAL       MINION        Weasel Tunneler    GANGS   
BRM_018                 PALADIN       MINION         Dragon Consort      BRM   
GVG_030                   DRUID       MINION      Anodized Robo Cub      GVG   
GVG_030                   DRUID       MINION      Anodized Robo Cub      GVG   
CS2_227                 NEUTRAL       MINION  Venture Co. Mercenary  EXPERT1   
KAR_062                 NEUTRAL       MINION  Netherspite Historian     KARA   
CFM_712_t17             NEUTRAL       MINION             Jade Golem    GANGS   
BRM_011t                 SHAMAN  ENCHANTMENT             Lava Shock      BRM   
BRMA10_3H               NEUTRAL   HERO_POWER            The Rookery      BRM   
KAR_710m                NEUTRAL       MINION        Animated Shield     KARA   
EX1_295o                   MAGE  ENCHANTMENT              Ice Block  EXPERT1   
OG_158                  NEUTRAL       MINION       Zealous Initiate       OG   
LOE_110t                NEUTRAL        SPELL          Ancient Curse      LOE   
LOE_110t                NEUTRAL        SPELL          Ancient Curse      LOE   
BRMA13_4H               NEUTRAL   HERO_POWER             Wild Magic      BRM   
EX1_145o                  ROGUE  ENCHANTMENT            Preparation  EXPERT1   
LOEA09_2e               NEUTRAL  ENCHANTMENT                Enraged      LOE   
EX1_570e                NEUTRAL  ENCHANTMENT                   Bite  EXPERT1   
GVG_069                 NEUTRAL       MINION        Antique Healbot      GVG   
hexfrog                 NEUTRAL       MINION                   Frog     CORE   
OG_145                  NEUTRAL       MINION           Psych-o-Tron       OG   
OG_145                  NEUTRAL       MINION           Psych-o-Tron       OG   
AT_090                  NEUTRAL       MINION       Mukla's Champion      TGT   
AT_060                   HUNTER        SPELL              Bear Trap      TGT   
TB_MechWar_Minion1      NEUTRAL       MINION               Mech Fan       TB   
...                         ...          ...                    ...      ...   
EX1_284                 NEUTRAL       MINION            Azure Drake  EXPERT1   
LOE_038                 NEUTRAL       MINION         Naga Sea Witch      LOE   
AT_083                  NEUTRAL       MINION       Dragonhawk Rider      TGT   
TB_CoOpv3_BOSS2e        NEUTRAL  ENCHANTMENT      Getting Angry....       TB   
GVG_009                  PRIEST       MINION           Shadowbomber      GVG   
CFM_338                  HUNTER       MINION       Trogg Beastrager    GANGS   
GVG_097                 NEUTRAL       MINION          Lil' Exorcist      GVG   
GVG_097                 NEUTRAL       MINION          Lil' Exorcist      GVG   
CFM_852                 NEUTRAL       MINION           Lotus Agents    GANGS   
LOE_020                  HUNTER       MINION           Desert Camel      LOE   
CFM_781                   ROGUE       MINION   Shaku, the Collector    GANGS   
GVG_090                 NEUTRAL       MINION          Madder Bomber      GVG   
GVG_120                 NEUTRAL       MINION       Hemet Nesingwary      GVG   
EX1_055o                NEUTRAL  ENCHANTMENT              Empowered  EXPERT1   
CFM_321                 NEUTRAL       MINION  Grimestreet Informant    GANGS   
EX1_082                 NEUTRAL       MINION             Mad Bomber  EXPERT1   
FP1_017                 NEUTRAL       MINION       Nerub'ar Weblord     NAXX   
FP1_001            

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads you displayed earlier.  Notice how it this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, conntains only 1079 rows.  Why?

Write your answer below this line:
________________________________________________________________________________

The reason why the `cards_with_mechanics_df` only contai

First performed an inner join, which only includes records that are present in both tables.  Although there were 2819 records in the left table, there were only 1079 records that existed in both tables, which are what you see in the resulting dataframe. 

## Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [ ]:
left_join_df = None
left_join_df

Note that the results of this sort of join are dependent upon the position of each table--if you were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________

Every record from `cards_with_mechanics_df`, as well as any records from `play_requirements_df` that have matching index values with a record from the left table.  

#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [ ]:
outer_join_df = None
outer_join_df

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, you learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)